OS X 10.5, PowerMac and "ERROR: Unknown or unidentifiable processor "powermacintosh""

Testing on old PowerMac; its circa 2005 or 2006. It has a 2.0 GHz PowerPC cpu, OS X 10.5.8, and MacPorts with updated gear. I keep it around for big-endian testing. It still runs faster than Core2 Duos, ARM gadgets and even Celerons.

Running `configure.py` produces the following.

```
PowerMac:botan$ /opt/local/bin/python2.7 ./configure.py --cc=clang --cc-bin=/opt/local/bin/clang++-mp-3.4 
   INFO: ./configure.py invoked with options "--cc=clang --cc-bin=/opt/local/bin/clang++-mp-3.4"
   INFO: Platform: OS="Darwin" machine="Power Macintosh" proc="powerpc"
   INFO: Guessing target OS is darwin (use --os to set)
  ERROR: Unknown or unidentifiable processor "powermacintosh"
```

This may be of interest:

```
PowerMac:botan$ uname -a
Darwin PowerMac.local 9.8.0 Darwin Kernel Version 9.8.0: Wed Jul 15 16:57:01 PDT 2009; root:xnu-1228.15.4~1/RELEASE_PPC Power Macintosh
PowerMac:botan$ uname -m
Power Macintosh
```

My apologies if this is unsupported. I saw there was some inline ASM code in the entropy gatherer to read a PowerPC's timestamp registers, so I thought it was game on.

It looks like this can be worked around by adding `--os=darwin --cpu=ppc64`. However, it looks line a variant of [Issue 528](https://github.com/randombit/botan/issues/528) resurfaced:

```
$ make
...
/opt/local/bin/clang++-mp-3.4  -m64 -pthread -stdlib=libc++ -fPIC -fvisibility=hidden -std=c++11 -D_REENTRANT -fstack-protector -O3 -Wall -Wextra -Wpedantic -Wshadow -Wstrict-aliasing -Wstrict-overflow=5 -Wcast-align -Wmissing-declarations -Wpointer-arith -Wcast-qual -Wunreachable-code -Ibuild/include -c ./src/lib/entropy/darwin_secrandom/darwin_secrandom.cpp -o build/obj/lib/entropy_darwin_secrandom.o
./src/lib/entropy/darwin_secrandom/darwin_secrandom.cpp:10:10: fatal error: 
      'Security/SecRandom.h' file not found
#include <Security/SecRandom.h>
         ^
1 error generated.
make: *** [build/obj/lib/entropy_darwin_secrandom.o] Error 1
```

--- a/configure.py
+++ b/configure.py
@@ -1165,11 +1165,11 @@ def system_cpu_info():
     except IOError:
         pass
 
-    if platform.processor() != '':
-        cpu_info.append(platform.processor())
-
     if platform.machine() != '':
         cpu_info.append(platform.machine())
+
+    if platform.processor() != '':
+        cpu_info.append(platform.processor())
 
     return cpu_info
 

--- a/configure.py
+++ b/configure.py
@@ -1151,15 +1151,19 @@ def system_cpu_info():
 
     cpu_info = []
 
-    with open('/proc/cpuinfo') as f:
-        for line in f.readlines():
-            if line.find(':') != -1:
-                (key,val) = [s.strip() for s in line.split(':')]
+    try:
+        with open('/proc/cpuinfo') as f:
+            for line in f.readlines():
+                if line.find(':') != -1:
+                    (key,val) = [s.strip() for s in line.split(':')]
 
-                # Different Linux arch use different names for this field in cpuinfo
-                if key in ["model name", "cpu model", "Processor"]:
-                    cpu_info.append(val)
-                    break
+                    # Different Linux arch use different names for this field in cpuinfo
+                    if key in ["model name", "cpu model", "Processor"]:
+                        cpu_info.append(val)
+                        break
+
+    except IOError:
+        pass
 
     if platform.processor() != '':
         cpu_info.append(platform.processor())

--- a/configure.py
+++ b/configure.py
@@ -1147,32 +1147,43 @@ def canon_processor(archinfo, proc):
                     proc, match, submodel))
                 return (ainfo.basename, submodel)
 
-    logging.debug('Known CPU names: ' + ' '.join(
-        sorted(flatten([[ainfo.basename] + \
-                        ainfo.aliases + \
-                        [x for (x,_) in ainfo.all_submodels()]
-                        for ainfo in archinfo.values()]))))
+def system_cpu_info():
 
-    raise Exception('Unknown or unidentifiable processor "%s"' % (proc))
+    cpu_info = []
+
+    with open('/proc/cpuinfo') as f:
+        for line in f.readlines():
+            if line.find(':') != -1:
+                (key,val) = [s.strip() for s in line.split(':')]
+
+                # Different Linux arch use different names for this field in cpuinfo
+                if key in ["model name", "cpu model", "Processor"]:
+                    cpu_info.append(val)
+                    break
+
+    if platform.processor() != '':
+        cpu_info.append(platform.processor())
+
+    if platform.machine() != '':
+        cpu_info.append(platform.machine())
+
+    return cpu_info
 
 def guess_processor(archinfo):
-    base_proc = platform.machine()
+    cpu_info = system_cpu_info()
 
-    if base_proc == '':
-        raise Exception('Could not determine target CPU; set with --cpu')
+    for input in cpu_info:
 
-    full_proc = fixup_proc_name(platform.processor()) or base_proc
+        if input != '':
+            try:
+                match = canon_processor(archinfo, input)
+                if match != None:
+                    logging.debug("Matched '%s' to processor '%s'" % (input, match))
+                    return match
+            except Exception as e:
+                logging.debug("Failed to deduce CPU from '%s'" % (input))
 
-    for ainfo in archinfo.values():
-        if ainfo.basename == base_proc or base_proc in ainfo.aliases:
-            for (match,submodel) in ainfo.all_submodels():
-                if re.search(match, full_proc) != None:
-                    return (ainfo.basename, submodel)
-
-            return canon_processor(archinfo, ainfo.basename)
-
-    # No matches, so just use the base proc type
-    return canon_processor(archinfo, base_proc)
+    raise Exception('Could not determine target CPU; set with --cpu')
 
 """
 Read a whole file into memory as a string
@@ -2108,6 +2119,12 @@ def main(argv = None):
     for policy in module_policies.values():
         policy.cross_check(modules)
 
+    logging.debug('Known CPU names: ' + ' '.join(
+        sorted(flatten([[ainfo.basename] + \
+                        ainfo.aliases + \
+                        [x for (x,_) in ainfo.all_submodels()]
+                        for ainfo in info_arch.values()]))))
+
     if options.list_modules:
         for k in sorted(modules.keys()):
             print(k)
@@ -2180,9 +2197,15 @@ def main(argv = None):
             options.arch, options.cpu))
     else:
         cpu_from_user = options.cpu
-        (options.arch, options.cpu) = canon_processor(info_arch, options.cpu)
-        logging.info('Canonicalizized CPU target %s to %s/%s' % (
-            cpu_from_user, options.arch, options.cpu))
+
+        results = canon_processor(info_arch, options.cpu)
+
+        if results != None:
+            (options.arch, options.cpu) = results
+            logging.info('Canonicalizized CPU target %s to %s/%s' % (
+                cpu_from_user, options.arch, options.cpu))
+        else:
+            logging.error('Unknown or unidentifiable processor "%s"' % (options.cpu))
 
     logging.info('Target is %s-%s-%s-%s' % (
         options.compiler, options.os, options.arch, options.cpu))

